src/Framework/PaymentGateways/PaymentGatewayRegister.php

```diff
@@ -99,20 +100,21 @@ public function unregisterGateway( $gatewayId ) {
 	}
 
 	/**
+	 *
+	 * Register Gateway with Service Container as Singleton
+	 * with option of adding Subscription Module through filter "give_gateway_{$gatewayId}_subscription_module"
+	 *
 	 * @unreleased
 	 *
 	 * @param  string  $gatewayClass
 	 *
 	 * @return void
 	 */
-	private function registerGatewayWithServiceContainer( $gatewayClass ) {
-		give()->singleton( $gatewayClass, function () use ( $gatewayClass ) {
-			/** @var PaymentGateway $gateway */
-			$gateway = new $gatewayClass;
-
-			do_action( "give_gateway_{$gateway->getId()}_mount_subscription_module", $gateway );
+	private function registerGatewayWithServiceContainer( $gatewayClass, $gatewayId ) {
+		give()->singleton( $gatewayClass, function ( Container $container ) use ( $gatewayClass, $gatewayId ) {
+			$subscriptionModule = apply_filters( "give_gateway_{$gatewayId}_subscription_module", null );
 
-			return $gateway;
+			return new $gatewayClass( $subscriptionModule ? $container->make( $subscriptionModule ) : null );
```

Comment:
Does this work?
```php
$parameters = $subscriptionModule ? [$container->make($subscriptionModule)] : null;
$return $container->make($gatewayClass, $parameters);
```

If we don't do that then automatic DI won't work for gateways in the future, and I can see that being useful.

Reply:
No did you see my comment about this? https://github.com/impress-org/givewp/pull/6071#discussion_r743743701

Reply:
Yeah, but I don't like it. It doesn't make any sense. Hahah! I'd like to figure out what's going on and see if we can pull this off.

Reply:
From my debugging I can see that the singleton callback is stuck in a loop.  When the container tries to resolve it keeps going to build the concrete and loops back around from here:

![Screen Shot 2021-11-05 at 1 18 00 PM](https://user-images.githubusercontent.com/10138447/140553634-0f4400c7-3345-42fb-a09e-5d4e15167896.png)

Reply:
I see. I'm guessing since we're using an older version of the Laravel Container this an issue. Let's just use `new` for now and see if it becomes and issue later.